<a href="https://colab.research.google.com/github/nurimammasri/Wooky-Pandas/blob/master/08.%20Grouping%20Data.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Introduction**

`Grouping` adalah suatu teknik memisahkan data berdasarkan kriteria tertentu. untuk melakukan grouping di pandas kita menggunakan suatu method yang bernama groupby(). definisi abstraknya dari grouping adalah kita melakukan mapping suatu data terhadap suatu group.

ketika melakukan grouping ada beberapa proses yang terjadi secara berurutan, yaitu :

1. **Splitting** : Memisahkan data kedalam suatu group berdasarkan kriteria tertentu.
2. **Applying** : Melakukan suatu operasi terhadap sekumpulan data di group-group tersebut.
3. **Combining** : Menggabungkan data menjadi suatu struktur baru

```
df.groupby(by=None, axis=0, level=None, as_index: bool=True, sort: bool=True, group_keys: bool=True, squeeze: bool=False, observed: bool=False)`
```
## **Parameters**

* `by` : mapping, function, label, or list of labels

    Used to determine the groups for the groupby.

    If ``by`` is a function, it's called on each value of the object's index. 
    
    If a dict or Series is passed, the Series or dict VALUES will be used to determine the groups (the Series' values are first
aligned; see ``.align()`` method). 

    If an ndarray is passed, the values are used as-is determine the groups. A label or list of labels may be passed to group by the columns in ``self``. 
    
    Notice that a tuple is interpreted as a (single) key.

* `axis` : {0 or 'index', 1 or 'columns'}, default 0

    Split along rows (0) or columns (1).

* `level` : int, level name, or sequence of such, default None

    If the axis is a MultiIndex (hierarchical), group by a particular level or levels.

* `as_index` : bool, default True

    For aggregated output, return object with group labels as the index. Only relevant for DataFrame input. as_index=False is effectively "SQL-style" grouped output.

* `sort` : bool, default True

    Sort group keys. Get better performance by turning this off. Note this does not influence the order of observations within each group. Groupby preserves the order of rows within each group.

* `group_keys` : bool, default True

    When calling apply, add group keys to index to identify pieces.

* `squeeze` : bool, default False

    Reduce the dimensionality of the return type if possible, otherwise return a consistent type.

* `observed` : bool, default False

    This only applies if any of the groupers are Categoricals.

    If True: only show observed values for categorical groupers.

    If False: show all values for categorical groupers.

```
df.groupby('class')

df.groupby('class').mean()

df.groupby(['class1','class2']).mean()

df.groupby('class').agg('mean')

df.groupby(['class1','class2'])[['column1', 'column2']].agg(['mean', 'sum', 'std'])

-----

def diff_min_max(x):
  return x.max()-x.min()

df.groupby(['class1','class2'])[['column1', 'column2']].agg(['sum', diff_min_max])

----

df.groupby(['class1','class2'])[['column1', 'column2']].agg({'column1':'sum', 'column2':'mean'})

```

In [1]:
import pandas as pd
df = pd.DataFrame()

In [2]:
df.groupby?

## Grouping Data

In [3]:
import numpy as np
import pandas as pd
df = pd.DataFrame([('bird', 'Falconiformes', 389.0),
                           ('bird', 'Psittaciformes', 24.0),
                           ('mammal', 'Carnivora', 80.2),
                           ('mammal', 'Primates', np.nan),
                           ('mammal', 'Carnivora', 58)],
                          index=['falcon', 'parrot', 'lion', 'monkey', 'leopard'],
                             columns=('class', 'order', 'max_speed'))

df

class           order  max_speed
falcon     bird   Falconiformes      389.0
parrot     bird  Psittaciformes       24.0
lion     mammal       Carnivora       80.2
monkey   mammal        Primates        NaN
leopard  mammal       Carnivora       58.0

In [4]:
df.groupby('class')

Coba kalian perhatikan, pada tahap ini kita baru sampai pada process **splitting** berdasarkan kategori ‘class’. hasil dari groupby ini hanya sebuah objek GroupBy. 

objek groupby ini adalah representasi dari suatu hasil penggabungan data berdasarkan group class, namun tidak menghasilkan nilai apapun sebelum kita memberikan suatu operasi terhadap hasil dari setiap grouping tersebut. 

prosess melakukan operasi terhadap sekumpulan data yang telah di groupkan adalah process **applying**. kita bisa menggunakan berbagai jenis operasi terhadap objek groupby tersebut. 

ada operasi yang telah di sediakan sebagai operasi bawaan dari pandas seperti mean(), min(), max(), dan lain sebagainya. 

dapat pula menggunakan custom operasi tergantung apa yang kita inginkan. mari kita coba praktekan salah satu jenis operasi kepada objek groupby tersebut.

In [5]:
df.groupby('class').mean()

max_speed
class            
bird        206.5
mammal       69.1

Coba kalian perhatikan. kita telah berhasil menggunakan groupby dan memberikan hasil. jenis operasi yang di apply terhadap setiap group adalah rata-rata. 

dan hasilnya di **combine** menjadi struktur data baru. dengan begini kita menhasilkan suatu hasil analisis sederhana yaitu bahwa binatang dengan class bird memiliki kecepatan maksimal rata-rata sebesar 206.5, dan bintang class mamalia sebesar 69.1

....................................................................................................................................................................

In [6]:
import pandas as pd
df = pd.DataFrame({'Animal': ['Falcon', 'Falcon', 'Parrot', 'Parrot'],
                    'Max Speed': [380., 370., 24., 26.]})
df

Animal  Max Speed
0  Falcon      380.0
1  Falcon      370.0
2  Parrot       24.0
3  Parrot       26.0

In [7]:
df.groupby(['Animal']).mean()

Max Speed
Animal           
Falcon      375.0
Parrot       25.0

In [8]:
df.groupby('Animal').agg('mean')

Max Speed
Animal           
Falcon      375.0
Parrot       25.0

**Hierarchical Indexes**

We can groupby different levels of a hierarchical index
using the `level` parameter:

In [9]:
arrays = [['Falcon', 'Falcon', 'Parrot', 'Parrot'],
          ['Captive', 'Wild', 'Captive', 'Wild']]
index = pd.MultiIndex.from_arrays(arrays, names=('Animal', 'Type'))
df = pd.DataFrame({'Max Speed': [390., 350., 30., 20.]}, index=index)
df

Max Speed
Animal Type              
Falcon Captive      390.0
       Wild         350.0
Parrot Captive       30.0
       Wild          20.0

In [10]:
df.groupby(level=0).mean()

Max Speed
Animal           
Falcon      370.0
Parrot       25.0

In [11]:
df.groupby(level="Animal").mean()

Max Speed
Animal           
Falcon      370.0
Parrot       25.0

In [12]:
df.groupby(level="Type").mean()

Max Speed
Type              
Captive      210.0
Wild         185.0

In [13]:
import pandas as pd
penjualan = {
    'Hari':['Hari-1']*6+['Hari-2']*6,
    'toko':['Bandung', 'Garut', 'Jakarta', 'Garut', 'Bandung', 'Jakarta', 'Bandung', 'Garut', 'Jakarta', 'Garut', 'Bandung', 'Jakarta'],
    'Barang':['Telur', 'Telur', 'Telur', 'Minyak', 'Minyak', 'Minyak', 'Telur', 'Telur', 'Telur', 'Minyak', 'Minyak', 'Minyak'],
    'Pendapatan_Kotor':[200000, 300000, 150000, 100000, 120000, 400000, 240000, 320000, 200000, 300000, 150000, 200000],
    'Pendapatan_Bersih':[80000, 100000, 50000, 50000, 50000, 200000, 120000, 150000, 120000, 100000, 110000, 90000]
}
df = pd.DataFrame(penjualan)
df

Hari     toko  Barang  Pendapatan_Kotor  Pendapatan_Bersih
0   Hari-1  Bandung   Telur            200000              80000
1   Hari-1    Garut   Telur            300000             100000
2   Hari-1  Jakarta   Telur            150000              50000
3   Hari-1    Garut  Minyak            100000              50000
4   Hari-1  Bandung  Minyak            120000              50000
5   Hari-1  Jakarta  Minyak            400000             200000
6   Hari-2  Bandung   Telur            240000             120000
7   Hari-2    Garut   Telur            320000             150000
8   Hari-2  Jakarta   Telur            200000             120000
9   Hari-2    Garut  Minyak            300000             100000
10  Hari-2  Bandung  Minyak            150000             110000
11  Hari-2  Jakarta  Minyak            200000              90000

In [14]:
df.groupby('Hari').sum()

Pendapatan_Kotor  Pendapatan_Bersih
Hari                                       
Hari-1           1270000             530000
Hari-2           1410000             690000

In [15]:
df.groupby(['Hari', 'toko', 'Barang']).sum()

Pendapatan_Kotor  Pendapatan_Bersih
Hari   toko    Barang                                     
Hari-1 Bandung Minyak            120000              50000
               Telur             200000              80000
       Garut   Minyak            100000              50000
               Telur             300000             100000
       Jakarta Minyak            400000             200000
               Telur             150000              50000
Hari-2 Bandung Minyak            150000             110000
               Telur             240000             120000
       Garut   Minyak            300000             100000
               Telur             320000             150000
       Jakarta Minyak            200000              90000
               Telur             200000             120000

## **Group by Multiple Columns**

In [16]:
df

Hari     toko  Barang  Pendapatan_Kotor  Pendapatan_Bersih
0   Hari-1  Bandung   Telur            200000              80000
1   Hari-1    Garut   Telur            300000             100000
2   Hari-1  Jakarta   Telur            150000              50000
3   Hari-1    Garut  Minyak            100000              50000
4   Hari-1  Bandung  Minyak            120000              50000
5   Hari-1  Jakarta  Minyak            400000             200000
6   Hari-2  Bandung   Telur            240000             120000
7   Hari-2    Garut   Telur            320000             150000
8   Hari-2  Jakarta   Telur            200000             120000
9   Hari-2    Garut  Minyak            300000             100000
10  Hari-2  Bandung  Minyak            150000             110000
11  Hari-2  Jakarta  Minyak            200000              90000

In [17]:
df.groupby(['Hari', 'toko'])[['Pendapatan_Kotor']].mean()

Pendapatan_Kotor
Hari   toko                     
Hari-1 Bandung            160000
       Garut              200000
       Jakarta            275000
Hari-2 Bandung            195000
       Garut              310000
       Jakarta            200000

In [18]:
df.groupby(['Hari', 'toko'])[['Pendapatan_Kotor', 'Pendapatan_Bersih']].mean()

Pendapatan_Kotor  Pendapatan_Bersih
Hari   toko                                        
Hari-1 Bandung            160000              65000
       Garut              200000              75000
       Jakarta            275000             125000
Hari-2 Bandung            195000             115000
       Garut              310000             125000
       Jakarta            200000             105000

## **Group by Multiple Aggregation**

In [19]:
df

Hari     toko  Barang  Pendapatan_Kotor  Pendapatan_Bersih
0   Hari-1  Bandung   Telur            200000              80000
1   Hari-1    Garut   Telur            300000             100000
2   Hari-1  Jakarta   Telur            150000              50000
3   Hari-1    Garut  Minyak            100000              50000
4   Hari-1  Bandung  Minyak            120000              50000
5   Hari-1  Jakarta  Minyak            400000             200000
6   Hari-2  Bandung   Telur            240000             120000
7   Hari-2    Garut   Telur            320000             150000
8   Hari-2  Jakarta   Telur            200000             120000
9   Hari-2    Garut  Minyak            300000             100000
10  Hari-2  Bandung  Minyak            150000             110000
11  Hari-2  Jakarta  Minyak            200000              90000

In [20]:
df.groupby(['Hari', 'toko'])[['Pendapatan_Kotor', 'Pendapatan_Bersih']].agg(['mean', 'sum', 'std'])

Pendapatan_Kotor          ... Pendapatan_Bersih               
                           mean     sum  ...               sum            std
Hari   toko                              ...                                 
Hari-1 Bandung           160000  320000  ...            130000   21213.203436
       Garut             200000  400000  ...            150000   35355.339059
       Jakarta           275000  550000  ...            250000  106066.017178
Hari-2 Bandung           195000  390000  ...            230000    7071.067812
       Garut             310000  620000  ...            250000   35355.339059
       Jakarta           200000  400000  ...            210000   21213.203436

[6 rows x 6 columns]

## **Group by with Custom Function**

In [21]:
df

Hari     toko  Barang  Pendapatan_Kotor  Pendapatan_Bersih
0   Hari-1  Bandung   Telur            200000              80000
1   Hari-1    Garut   Telur            300000             100000
2   Hari-1  Jakarta   Telur            150000              50000
3   Hari-1    Garut  Minyak            100000              50000
4   Hari-1  Bandung  Minyak            120000              50000
5   Hari-1  Jakarta  Minyak            400000             200000
6   Hari-2  Bandung   Telur            240000             120000
7   Hari-2    Garut   Telur            320000             150000
8   Hari-2  Jakarta   Telur            200000             120000
9   Hari-2    Garut  Minyak            300000             100000
10  Hari-2  Bandung  Minyak            150000             110000
11  Hari-2  Jakarta  Minyak            200000              90000

In [22]:
def diff_min_max(x):
  return x.max()-x.min()

df.groupby(['Hari', 'toko'])[['Pendapatan_Kotor', 'Pendapatan_Bersih']].agg(['sum', diff_min_max])

Pendapatan_Kotor              Pendapatan_Bersih             
                            sum diff_min_max               sum diff_min_max
Hari   toko                                                                
Hari-1 Bandung           320000        80000            130000        30000
       Garut             400000       200000            150000        50000
       Jakarta           550000       250000            250000       150000
Hari-2 Bandung           390000        90000            230000        10000
       Garut             620000        20000            250000        50000
       Jakarta           400000            0            210000        30000

## **Custom Aggregation with Dictionary**

In [23]:
df.groupby(['Hari', 'toko'])[['Pendapatan_Kotor', 'Pendapatan_Bersih']].agg({'Pendapatan_Kotor':'sum', 'Pendapatan_Bersih':'mean'})

Pendapatan_Kotor  Pendapatan_Bersih
Hari   toko                                        
Hari-1 Bandung            320000              65000
       Garut              400000              75000
       Jakarta            550000             125000
Hari-2 Bandung            390000             115000
       Garut              620000             125000
       Jakarta            400000             105000

In [24]:
df.groupby(['Hari', 'toko'])[['Pendapatan_Kotor', 'Pendapatan_Bersih']].agg({'Pendapatan_Kotor':'sum', 'Pendapatan_Bersih':'mean'})

Pendapatan_Kotor  Pendapatan_Bersih
Hari   toko                                        
Hari-1 Bandung            320000              65000
       Garut              400000              75000
       Jakarta            550000             125000
Hari-2 Bandung            390000             115000
       Garut              620000             125000
       Jakarta            400000             105000

## **Task**

In [25]:
# '/content/drive/My Drive/Colab Notebooks/Wooky Pandas/diamonds.csv'
# import pandas as pd
# df = pd.read_csv('https://drive.google.com/uc?id=1NShkvOY-bxIulZZ8iWczg0-f8g3l1QvI')
# df

In [26]:
# '/content/drive/My Drive/Colab Notebooks/Wooky Pandas/diamonds.csv'
import pandas as pd
df = pd.read_csv('https://raw.githubusercontent.com/nurimammasri/Wooky-Pandas/master/Data/diamonds.csv')
df

Unnamed: 0  carat        cut color  ... price     x     y     z
0               1   0.23      Ideal     E  ...   326  3.95  3.98  2.43
1               2   0.21    Premium     E  ...   326  3.89  3.84  2.31
2               3   0.23       Good     E  ...   327  4.05  4.07  2.31
3               4   0.29    Premium     I  ...   334  4.20  4.23  2.63
4               5   0.31       Good     J  ...   335  4.34  4.35  2.75
...           ...    ...        ...   ...  ...   ...   ...   ...   ...
53935       53936   0.72      Ideal     D  ...  2757  5.75  5.76  3.50
53936       53937   0.72       Good     D  ...  2757  5.69  5.75  3.61
53937       53938   0.70  Very Good     D  ...  2757  5.66  5.68  3.56
53938       53939   0.86    Premium     H  ...  2757  6.15  6.12  3.74
53939       53940   0.75      Ideal     D  ...  2757  5.83  5.87  3.64

[53940 rows x 11 columns]

In [27]:
df.head()

Unnamed: 0  carat      cut color clarity  ...  table  price     x     y     z
0           1   0.23    Ideal     E     SI2  ...   55.0    326  3.95  3.98  2.43
1           2   0.21  Premium     E     SI1  ...   61.0    326  3.89  3.84  2.31
2           3   0.23     Good     E     VS1  ...   65.0    327  4.05  4.07  2.31
3           4   0.29  Premium     I     VS2  ...   58.0    334  4.20  4.23  2.63
4           5   0.31     Good     J     SI2  ...   58.0    335  4.34  4.35  2.75

[5 rows x 11 columns]

**Columns Metadata:**

`carat`: berat diamond

`cut`: kualitas diamond

`color`: warna diamond

`clarity`: level kemurnian diamonds

`depth`: tinggi dari diamond tabel

`table`: lebar  dari diamond tabel

`price`: harga diamond ($US)

`x`: panjang

`y`: lebar

`z`: kedalaaman

<h2>Soal 1: Process Grouping</h2>

Jelaskan prosess yang terjadi dalam grouping data menggungkan method <b>groupby()</b> dari pandas.

jawab:

Grouping adalah suatu teknik memisahkan data berdasarkan kriteria tertentu. untuk melakukan grouping di pandas kita menggunakan suatu method yang bernama groupby(). definisi abstraknya dari grouping adalah kita melakukan mapping suatu data terhadap suatu group.

ketika melakukan grouping ada beberapa proses yang terjadi secara berurutan, yaitu :

Splitting, Applying, Combining

pada Grouping kita juga bisa membuat suatu fungsi Aggregation seperti SUM, MIN, MAX, AVG

<h2>Soal 2: Grouping All Column</h2>
    
Group data diamond berdasarkan cut (kualitas), lakukan aggregasi dengan fungsi rata2.

Kemudian jawab pertanyaan berikut.
- Berapa harga rata2 dari diamond berkualitas Premium

Expected Output:

![grouping_data01.png](https://raw.githubusercontent.com/nurimammasri/Wooky-Pandas/master/Images/grouping_data01.png)

In [28]:
# code 
df.groupby('cut').agg('mean')

Unnamed: 0     carat      depth  ...         x         y         z
cut                                           ...                              
Fair       24147.357764  1.046137  64.041677  ...  6.246894  6.182652  3.982770
Good       24774.931309  0.849185  62.365879  ...  5.838785  5.850744  3.639507
Ideal      29047.630736  0.702837  61.709401  ...  5.507451  5.520080  3.401448
Premium    25600.209049  0.891955  61.264673  ...  5.973887  5.944879  3.647124
Very Good  26097.313193  0.806381  61.818275  ...  5.740696  5.770026  3.559801

[5 rows x 8 columns]

In [29]:
df.groupby('cut').mean().loc['Premium','price']

4584.2577042999055

jawab: **4584.2577042999055**

<h2>Soal 3: Grouping Some Column base on Multiple Criteria</h2>
    
Lakukan groupby pada data berdasarkan kriteria cut (kualitas) dan color (warna) secara berurutan, dengan aggregasi pada column carat, table, dan price saja. dengan fungsi agregasi rata-rata.

Kemudian jawab pertanyaan berikut:
- Berapa harga diamond yang berkriteria premium dan berwarna J
- Berapa nilai dari carat diamond yang berkriteria very good dan berwarna H

Expected Output:

Note: gambar hanya sebagaian dari keseluruhan data

![grouping_data02.png](https://raw.githubusercontent.com/nurimammasri/Wooky-Pandas/master/Images/grouping_data02.png)

In [30]:
# code here
data = df.groupby(['cut','color'])[['carat', 'table', 'price']].mean()
data

carat      table        price
cut       color                                  
Fair      D      0.920123  58.969325  4291.061350
          E      0.856607  59.364732  3682.312500
          F      0.904712  59.453205  3827.003205
          G      1.023822  58.773248  4239.254777
          H      1.219175  58.696370  5135.683168
          I      1.198057  59.237143  4685.445714
          J      1.341176  58.917647  4975.655462
Good      D      0.744517  58.541541  3405.382175
          E      0.745134  58.779957  3423.644159
          F      0.775930  58.910891  3495.750275
          G      0.850896  58.471986  4123.482204
          H      0.914729  58.611111  4276.254986
          I      1.057222  58.773946  5078.532567
          J      1.099544  58.813029  4574.172638
Ideal     D      0.565766  55.965632  2629.094566
          E      0.578401  55.967461  2597.550090
          F      0.655829  55.924203  3374.939362
          G      0.700715  55.902375  3720.706388
          H      0.799525  55.965843  3889.334831
          I      0.913029  56.021357  4451.970377
          J      1.063594  56.012612  4918.186384
Premium   D      0.721547  58.718964  3631.292576
          E      0.717745  58.779461  3538.914420
          F      0.827036  58.679279  4324.890176
          G      0.841488  58.702360  4500.742134
          H      1.016449  58.792034  5216.706780
          I      1.144937  58.771849  5946.180672
          J      1.293094  58.874752  6294.591584
Very Good D      0.696424  58.041309  3470.467284
          E      0.676317  58.038875  3214.652083
          F      0.740961  57.848429  3778.820240
          G      0.766799  57.784428  3872.753806
          H      0.915948  57.903015  4535.390351
          I      1.046952  58.105150  5255.879568
          J      1.133215  58.277729  5103.513274

In [31]:
# harga diamond yang berkriteria premium dan berwarna J
data.loc['Premium'].loc['J', 'price']

6294.591584158416

In [32]:
# nilai dari carat diamond yang berkriteria very good dan berwarna H
data.loc['Very Good'].loc['H', 'carat']

0.9159484649122761

jawab:
*   Berapa harga diamond yang berkriteria premium dan berwarna J = 6294.591584158416
*   Berapa nilai dari carat diamond yang berkriteria very good dan berwarna H = 0.9159484649122761





<h2>Soal 4: Group By Multiple Aggregate Base on Multiple Criteria</h2>
    
Lakukan groupby pada data berdasarkan kriteria cut(kualitas) dan color (warna) secara berurutan kemudian aggregasi objek group by nya dengan fungsi rata2 dan median.

Kemudian Jawab Pertanyaan berikut:

- Berapa nilai median dari harga diamond berkualitas good dan berwarna F

Expected Output:

Note : gambar hanya sebagaian dari keseluruhan data

![grouping_data03.png](https://raw.githubusercontent.com/nurimammasri/Wooky-Pandas/master/Images/grouping_data03.png)

In [33]:
dataagg = df.groupby(['cut', 'color']).agg(['mean', 'median'])
dataagg

Unnamed: 0              carat  ...      y         z       
                         mean   median      mean  ... median      mean median
cut       color                                   ...                        
Fair      D      22993.699387  15921.0  0.920123  ...  6.010  3.839877  3.900
          E      24044.593750  19368.0  0.856607  ...  6.020  3.722143  3.760
          F      24414.756410  19651.5  0.904712  ...  5.990  3.787821  3.815
          G      26670.949045  24338.0  1.023822  ...  6.040  3.963153  3.970
          H      22296.056106  18202.0  1.219175  ...  6.250  4.219373  4.090
          I      24912.960000  20569.0  1.198057  ...  6.270  4.193486  4.080
          J      22148.983193  17416.0  1.341176  ...  6.460  4.319664  4.150
Good      D      23869.592145  19696.0  0.744517  ...  5.730  3.504864  3.540
          E      26368.867095  26054.0  0.745134  ...  5.710  3.496066  3.480
          F      25879.342134  24988.0  0.775930  ...  5.790  3.544609  3.580
          G      24284.672790  20908.0  0.850896  ...  6.070  3.647922  3.790
          H      23798.809117  21549.5  0.914729  ...  6.130  3.726937  3.830
          I      24939.764368  22865.5  1.057222  ...  6.260  3.901513  3.895
          J      21955.690554  18373.0  1.099544  ...  6.460  3.977459  4.010
Ideal     D      30917.312279  33756.0  0.565766  ...  5.110  3.201800  3.160
          E      30868.591852  34081.0  0.578401  ...  5.110  3.220346  3.150
          F      29760.244903  32699.5  0.655829  ...  5.240  3.338732  3.220
          G      29367.414619  30903.5  0.700715  ...  5.260  3.403120  3.230
          H      26684.482504  26903.0  0.799525  ...  5.710  3.539917  3.510
          I      26418.362637  25633.0  0.913029  ...  5.860  3.695270  3.600
          J      24773.177455  22691.5  1.063594  ...  6.510  3.912757  4.010
Premium   D      27467.155958  29439.0  0.721547  ...  5.340  3.415926  3.300
          E      28167.829696  30556.0  0.717745  ...  5.360  3.408875  3.290
          F      26299.272844  25039.0  0.827036  ...  5.860  3.588391  3.590
          G      26703.963406  25849.5  0.841488  ...  5.870  3.579641  3.610
          H      22654.713136  20799.0  1.016449  ...  6.450  3.811699  3.960
          I      22516.406162  21663.5  1.144937  ...  6.690  3.963894  4.130
          J      22512.214109  20841.5  1.293094  ...  6.900  4.168020  4.260
Very Good D      25980.308658  26210.0  0.696424  ...  5.490  3.405089  3.360
          E      27405.045417  28624.0  0.676317  ...  5.365  3.373996  3.280
          F      26471.613216  25408.0  0.740961  ...  5.720  3.472902  3.530
          G      26984.105263  26413.0  0.766799  ...  5.710  3.509883  3.540
          H      24184.660088  22198.5  0.915948  ...  6.160  3.713849  3.830
          I      25298.674419  22629.0  1.046952  ...  6.360  3.890972  3.935
          J      24091.399705  21462.0  1.133215  ...  6.575  4.006873  4.060

[35 rows x 16 columns]

In [34]:
# Berapa nilai median dari harga diamond berkualitas good dan berwarna F
dataagg.loc['Good', 'price'].loc['F', 'median']


2647.0

jawab:

2647.0

<h2>Soal 5: Group By With Different Aggregation for specific Column</h2>
    
Lakukan groupby pada data berdasarkan criteria cut(kualitas), kemudian aggreagate dengan fungsi max pada column price dan fungsi min pada column carat.

Kemudian Jawab Pertanyaan berikut:

- Berapa nilai max dari price untuk diamond berkualitas Ideal

Expected Output:

![grouping_data04.png](https://raw.githubusercontent.com/nurimammasri/Wooky-Pandas/master/Images/grouping_data04.png)

In [35]:
# code here
dataagg2 = df.groupby('cut')[['price', 'carat']].agg({'price':'max', 'carat':'min'})
dataagg2

price  carat
cut                    
Fair       18574   0.22
Good       18788   0.23
Ideal      18806   0.20
Premium    18823   0.20
Very Good  18818   0.20

In [36]:
# Berapa nilai max dari price untuk diamond berkualitas Ideal
dataagg2.loc['Ideal', 'price']

18806

jawab:

18806

<h2>Soal 6: Group By With Custom Function</h2>
    
Lakukan groupby pada data berdasarkan criteria cut(kualitas), kemudian lakukan aggregasi dengan fungsi standar deviasi pada column carat dan fungsi rentang nilai antara max dan min pada column price.

Kemudian jawab pertanyaan berikut:
- Berapa selisih harga tertinggi dan terendah dari diamond berkualitas Premium

Expected Output:

![grouping_data05.png](https://raw.githubusercontent.com/nurimammasri/Wooky-Pandas/master/Images/grouping_data05.png)

In [37]:
# code here
def intrvl(x) :
  return x.max() - x.min()

dataagg3 = df.groupby('cut')[['price', 'carat']].agg({'price': intrvl,'carat':'std'})
dataagg3

price     carat
cut                       
Fair       18237  0.516404
Good       18461  0.454054
Ideal      18480  0.432876
Premium    18497  0.515262
Very Good  18482  0.459435

In [38]:
# Berapa selisih harga tertinggi dan terendah dari diamond berkualitas Premium
dataagg3.loc['Premium', 'price']

18497

jawab:

18497